esp/src/src-react/components/Metrics.tsx

```diff
@@ -0,0 +1,271 @@
+    filter?: object;
+}
+
+const emptyFilter = {};
+
+export const Metrics: React.FunctionComponent<MetricsProps> = ({
+    wuid,
+    filter = emptyFilter
+}) => {
+    const [_uiState, _setUIState] = React.useState({ ...defaultUIState });
+    const [timelineFilter, setTimelineFilter] = React.useState("");
+    const [metrics, _columns, _activities, _properties, _measures, _scopeTypes] = useWorkunitMetrics(wuid);
+    const [showMetricOptions, setShowMetricOptions] = React.useState(false);
+    const [options] = useMetricsOptions();
+    const [isFavorite, addFavorite, removeFavorite] = useFavorite(window.location.hash);
+
+    //  Command Bar  ---
+    const buttons = React.useMemo((): ICommandBarItemProps[] => [
+        {
+            key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
+            onClick: () => { }
+        },
+        { key: "divider_1", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "options", text: nlsHPCC.Options, iconProps: { iconName: "Settings" },
+            onClick: () => {
+                setShowMetricOptions(true);
+            }
+        },
+    ], []);
+
+    const rightButtons = React.useMemo((): ICommandBarItemProps[] => [
+        {
+            key: "star", iconProps: { iconName: isFavorite ? "FavoriteStarFill" : "FavoriteStar" },
+            onClick: () => {
+                if (isFavorite) {
+                    removeFavorite();
+                } else {
+                    addFavorite();
+                }
+            }
+        },
+    ], [addFavorite, isFavorite, removeFavorite]);
+
+    //  Timeline ---
+    const timeline = useConst(() => new WUTimelinePatched()
+        .maxZoom(Number.MAX_SAFE_INTEGER)
+        .baseUrl("")
+        .request({
+            ScopeFilter: {
+                MaxDepth: 3,
+                ScopeTypes: []
+            },
+            NestedFilter: {
+                Depth: 0,
+                ScopeTypes: []
+            },
+            PropertiesToReturn: {
+                AllProperties: false,
+                AllStatistics: true,
+                AllHints: false,
+                Properties: ["WhenStarted", "TimeElapsed"]
+            },
+            ScopeOptions: {
+                IncludeId: true,
+                IncludeScope: true,
+                IncludeScopeType: true
+            },
+            PropertyOptions: {
+                IncludeName: true,
+                IncludeRawValue: true,
+                IncludeFormatted: true,
+                IncludeMeasure: true,
+                IncludeCreator: true,
+                IncludeCreatorType: false
+            }
+        })
+        .on("click", (row, col, sel) => {
+            setTimelineFilter(sel ? row[7].ScopeName : "");
+        })
+    );
+
+    React.useEffect(() => {
+        timeline.wuid(wuid);
+    }, [timeline, wuid]);
+
+    //  Scopes Table  ---
+    const hasFilter = Object.keys(filter).length > 0;
```

Comment:
useMemo()? I don't have a great handle on when to use (or not). I've only ever seen docs cite using it for "expensive" functions, which I can't imagine this would ever be. But it also just prevents undesirable re-renders?

Reply:
I hear you (and doesn't help that I have been somewhat inconsistent as I learn about these things) - Here is my take - there are two things to watch out for:
1.  If the function expensive (no)
2. Is the returned value (boolean in this case) expensive to hash (which is what React does under the covers to spot if something has actually changed) - so if the returned item was a large array or complex object then that could play a performance part as well.

So in this case there won't be any additional re-renders and its "lite".